### PR TITLE
Fix memory leak in `Oj.dump` if raise an exception

### PR DESCRIPTION
**Pull request by Watson1978**
If an exception occurs inside Oj.dump, the heap area has not been released.
The heap area allocated by ALLOC_N must be released.
This patch will fix this problem.

### Test code
```ruby
require 'oj'

data = {
  "foo" => "a" * 1024 * 1024 * 10,
  1 => "Invalid hash key with strict mode"
}

1000.times do |i|
  begin
    Oj.dump(data, mode: :strict)
  rescue
  end

  if i % 100 == 0
    rss = Integer(`ps -o rss= -p #{Process.pid}`) / 1024.0
    puts "#{i}, #{rss}"
  end
end
```

### Before
```
$ ruby json.rb
0, 49.4375
100, 1051.0
200, 1674.03125
300, 1773.625
400, 1822.171875
500, 1842.078125
600, 1841.015625
700, 1815.65625
800, 1781.53125
900, 1792.28125
```

1.8 GB of memory was used in my enviroment.

### After
```
$ ruby json.rb
0, 49.40625
100, 49.421875
200, 49.53125
300, 49.59375
400, 49.65625
500, 49.6875
600, 49.71875
700, 49.734375
800, 49.765625
900, 49.796875
```

50 MB of memory was used in my enviroment.

### CODE REVIEW

**Comment by Watson1978** [comment]
> Just one tiny request for a change.

OK, sure. What kind of request is that?

**Review by ohler55 — CHANGES_REQUESTED** [review]

**Comment by ohler55** [comment]
Sorry, I've gotten used to using gitlab and forgot to click the submit review.

File: ext/oj/oj.c [hunk 1]
```diff
@@ -1237,6 +1237,15 @@ static VALUE safe_load(VALUE self, VALUE doc) {
...
  */
 
+struct dump_arg {
+    struct _out     *out;
+    struct _options *copts;
+    int    argc;
+    VALUE *argv;
+};
+static VALUE dump_body(VALUE a);
+static VALUE dump_ensure(VALUE a);
```
[the remark this answers is not in the document]
Comment (Watson1978) [reply ext/oj/oj.c]: OK. Fixed at https://github.com/ohler55/oj/pull/676/commits/0ab66a6ffc05a336585ed15ff5713d53ac01e1de

Thanks

**Comment by ohler55** [comment]
Thank you for the fix.

**Comment by ohler55** [comment]
I don't have your email address but Oj is getting a parser rework and I would be honoured if you would like to act as a reviewer of the code when it is ready for a merge. Early result show the new parser could be as much as twice as fast as the current Oj parser.

**Comment by Watson1978** [comment]
Sounds great. OK, sure :)

**Pull request merged**